### PR TITLE
Hook up payment methods configuration

### LINKED ISSUE #4088: Settings Synchronization: Hook up payment methods configuration
We aim to allow merchants to update gateway settings from both WC Stripe settings and Stripe dashboard settings.

In https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3916 we are fetching the payment method configuration settings. For this issue, we want to hook up enabled payment methods(admin settings) to payment method configurations. After we are able to toggle settings from Stripe, and see those changes in the WooCommerce Stripe settings, we also want to be able to update the Stripe dashboard configuration settings from the WC Stripe settings.

[udpate_enabled_payment_methods](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/91353c5776f488a8fcbbb8347ea080942b06418c/includes/admin/class-wc-rest-stripe-settings-controller.php#L599-L643) should be updated to use the Stripe configuration we added in https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3916.

Image

Image

**Acceptance criteria**
- [ ] Merchants should be able to toggle payment gateway settings from Stripe dashboard settings and we should see those changes in the WooCommerce Stripe settings.
- [ ] Merchants should be able to toggle payment gateway settings from WC Stripe settings, and we should see those changes in the Stripe dashboard settings.

**Notes**
* Blocked by #3916 
* Create a PR based on the [feature branch](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4002)
* [Update payment method configuration](https://docs.stripe.com/api/payment_method_configurations/update)

### PR DESCRIPTION
**Pull request by hsingyuc**
Fixes #4088 

## Changes proposed in this Pull Request:
This PR hooks up payment method configuration settings to the admin Stripe enabled payment methods.

Out of scope:
* Disabling unsupported payment methods on Stripe.
* Google Pay / Apple Pay.
* Boleto and OXXO
* giropay and Sofort
* Track updating payment methods will be added in another PR.

https://github.com/user-attachments/assets/295715ab-a030-4b81-b1e3-a98a89312e1a

## Testing instructions
Test USD payment methods - 
Test EURO payment methods - Cards, 

**Amazon Pay as an example. Please also test other payment gateways and remember to switch to EURO**
1. Turn the feature flag on, by setting the `_wcstripe_feature_amazon_pay` option to yes
2. Verify that Amazon Pay is now available in Stripe settings
4. Set the currency to USD
5. Go to Stripe settings and enable Cards, Alipay, Amazon Pay, Cash app pay, Link, WeChat, Affirm, Afterpay, Klarna, and ACH Direct Debit.
6. All payment gateway settings from step 5 for the merchant copy configuration in the [Stripe dashboard your merchant copy ](https://dashboard.stripe.com/test/settings/payment_methods) should be enabled 
7. Disable All payment gateway settings from step 5 from Stripe dashboard settings
8. All payment gateway settings from step 5 should be disabled in admin Stripe settings
9. Set the currency to EURO
10. Go to Stripe settings and enable Link, Multibanco, Bancontact, EPS, iDEAL, Przelewy24, and SEPA Direct Debit
11. All payment gateway settings from step 10 for the merchant copy configuration in the [Stripe dashboard your merchant copy ](https://dashboard.stripe.com/test/settings/payment_methods) should be enabled 
12. Disable All payment gateway settings from step 10 from Stripe dashboard settings
8. All payment gateway settings from step 10 should be disabled in admin Stripe settings

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

### CODE REVIEW

File: includes/admin/class-wc-rest-stripe-settings-controller.php [hunk 1]
```diff
@@ -388,6 +414,29 @@ private function update_is_test_mode_enabled( WP_REST_Request $request ) {
...
 
+	/**
+	 * Get the merchant payment method configuration in Stripe.
+	 *
+	 * @return array|null
+	 */
+	private function get_merchant_payment_method_configuration() {
+		$payment_method_configurations = $this->stripe_api->get_payment_method_configurations() && property_exists( $this->stripe_api->get_payment_method_configurations(), 'data' )
+			? $this->stripe_api->get_payment_method_configurations()->data
+			: null;
```
Comment (annemirasol) [inline includes/admin/class-wc-rest-stripe-settings-controller.php]: What do we think about storing `$this->stripe_api->get_payment_method_configurations()` in an intermediate variable and doing the checks on that, to avoid multiple hits on the endpoint?

Something like
```suggestion
$result = $this->stripe_api->get_payment_method_configurations();
$payment_method_configurations = $result->data ?? null;
```

File: includes/admin/class-wc-rest-stripe-settings-controller.php [hunk 2]
```diff
@@ -388,6 +414,29 @@ private function update_is_test_mode_enabled( WP_REST_Request $request ) {
 		$this->gateway->update_option( 'testmode', $is_test_mode_enabled ? 'yes' : 'no' );
 	}
 
+	/**
+	 * Get the merchant payment method configuration in Stripe.
+	 *
+	 * @return array|null
+	 */
+	private function get_merchant_payment_method_configuration() {
+		$payment_method_configurations = $this->stripe_api->get_payment_method_configurations() && property_exists( $this->stripe_api->get_payment_method_configurations(), 'data' )
```
Comment (malithsen) [inline includes/admin/class-wc-rest-stripe-settings-controller.php]: In addition to what Anne mentioned below, calling the Stripe API every time the settings page is loaded will impact page load performance.

Is there a way to avoid that? 

One approach is to tie it up with the existing "Refresh account details" button and avoid automatic fetches.  
If configuration updates result in a webhook event, another approach might be to listen to those for updates.

File: includes/admin/class-wc-rest-stripe-settings-controller.php [hunk 3]
```diff
@@ -388,6 +414,29 @@ private function update_is_test_mode_enabled( WP_REST_Request $request ) {
 		$this->gateway->update_option( 'testmode', $is_test_mode_enabled ? 'yes' : 'no' );
 	}
 
+	/**
+	 * Get the merchant payment method configuration in Stripe.
+	 *
+	 * @return array|null
```
Comment (malithsen) [inline includes/admin/class-wc-rest-stripe-settings-controller.php]: Looks like this function returns an object instead of an array.

```suggestion
	 * @return object|null
```

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 2 above)
Comment (malithsen) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: On a separate note, we'd want to add a log if the call to Stripe API fails for some reason.

```
if ( is_wp_error( $response ) ) {
    WC_Stripe_Logger::log( 'Failed to fetch payment method configurations from Stripe API.' );
    return null;
}
```

**Review by malithsen — COMMENTED** [review]
Nice work so far here 👍 

Everything tests and works as expected.
✅ Updates to the payment methods on Stripe dashboard are reflected on the settings page. 
✅ Updates to the settings page are propagated to Stripe's configuration.
✅ Tested with USD/EUR.

One question: On my Stripe dashboard, there are two configurations.
1. "Your configuration"
2. "WooCommerce Inc. Configuration"

2nd config is what this PR uses. Is that the correct behavior?

Left some comments about page load performance which we might want to address before merge.

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 2 above)
Comment (hsingyuc) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: >One approach is to tie it up with the existing "Refresh account details" button and avoid automatic fetches.
If configuration updates result in a webhook event, another approach might be to listen to those for updates.

This could work, but this would change how the feature works and prevent users who update settings from the Stripe dashboard from seeing them immediately in admin settings. @diegocurbelo, would this work for us?

**Comment by hsingyuc** [comment]
>One question: On my Stripe dashboard, there are two configurations.
"Your configuration"
"WooCommerce Inc. Configuration"
2nd config is what this PR uses. Is that the correct behavior?

@diegocurbelo, WooCommerce Inc. configuration is the [merchant copy](https://a8c.slack.com/archives/C05GDK92WGM/p1741804587080839?thread_ts=1741751645.624259&cid=C05GDK92WGM) that is the one we should update, correct?

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 2 above)
Comment (hsingyuc) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: >On a separate note, we'd want to add a log if the call to Stripe API fails for some reason.

Yes, and we already have [a log here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/bc2bce601e372807d4e2d755f78a321d43330675/includes/class-wc-stripe-api.php/#L217-L220)

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 1 above)
Comment (hsingyuc) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: Thank you @annemirasol! Updated [ebb6e90](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4091/commits/ebb6e90885011cb846191dd4f0a727935676bd9e)

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 3 above)
Comment (hsingyuc) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: Thank you @malithsen ! Updated [ebb6e90](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4091/commits/ebb6e90885011cb846191dd4f0a727935676bd9e)

**Comment by annemirasol** [comment]
Is connecting the settings to the store front out of scope for this PR? For example, if I enable or disable Klarna in the settings page, the checkout page does not reflect the changes.

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 2 above)
Comment (diegocurbelo) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: > > One approach is to tie it up with the existing "Refresh account details" button and avoid automatic fetches.
> > If configuration updates result in a webhook event, another approach might be to listen to those for updates.
> 
> This could work, but this would change how the feature works and prevent users who update settings from the Stripe dashboard from seeing them immediately in admin settings. @diegocurbelo, would this work for us?

Using the "Refresh payment methods" in the options menu was discussed in the Figma 0fiyiUe18lGU6kGyQNmXxE-fi-165_4304 (or maybe in P2 🤷🏼), but as the settings page is not frequently visited and most of the time (if not always) the merchant will want up-to-date information it makes sense to fetch the enabled PMs every time. Using webhooks would be a valid alternative, but it adds another point of failure.

**Comment by diegocurbelo** [comment]
> > One question: On my Stripe dashboard, there are two configurations.
> > "Your configuration"
> > "WooCommerce Inc. Configuration"
> > 2nd config is what this PR uses. Is that the correct behavior?
> 
> @diegocurbelo, WooCommerce Inc. configuration is the p1741804587080839/1741751645.624259-slack-C05GDK92WGM that is the one we should update, correct?

Correct, that's the merchant-editable copy of the platform PMC.

File: includes/admin/class-wc-rest-stripe-settings-controller.php [hunk 4]
```diff
@@ -620,26 +631,7 @@ private function update_enabled_payment_methods( WP_REST_Request $request ) {
...
-			}
-		}
-
-		$this->gateway->update_option( 'upe_checkout_experience_accepted_payments', $upe_checkout_experience_accepted_payments );
-
-		// After updating payment methods record tracks events.
-		$newly_enabled_methods  = array_diff( $upe_checkout_experience_accepted_payments, $currently_enabled_payment_method_ids );
-		$newly_disabled_methods = array_diff( $currently_enabled_payment_method_ids, $payment_method_ids_to_enable );
-
-		$this->record_payment_method_settings_event( $newly_enabled_methods, $newly_disabled_methods );
```
Comment (Mayisha) [inline includes/admin/class-wc-rest-stripe-settings-controller.php:642]: Is this tracking part accidentally removed? Or is it intentionally removed in favor of https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3920?

File: includes/admin/class-wc-rest-stripe-settings-controller.php [hunk 5]
```diff
@@ -241,7 +248,7 @@ public function get_settings() {
 		$is_upe_enabled               = WC_Stripe_Feature_Flags::is_upe_checkout_enabled();
 		$available_payment_method_ids = $is_upe_enabled ? $this->gateway->get_upe_available_payment_methods() : WC_Stripe_Helper::get_legacy_available_payment_method_ids();
 		$ordered_payment_method_ids   = $is_upe_enabled ? WC_Stripe_Helper::get_upe_ordered_payment_method_ids( $this->gateway ) : $available_payment_method_ids;
-		$enabled_payment_method_ids   = $is_upe_enabled ? WC_Stripe_Helper::get_upe_settings_enabled_payment_method_ids( $this->gateway ) : WC_Stripe_Helper::get_legacy_enabled_payment_method_ids();
+		$enabled_payment_method_ids   = $is_upe_enabled ? $this->gateway->get_upe_enabled_payment_method_ids() : [];
```
Comment (Mayisha) [inline includes/admin/class-wc-rest-stripe-settings-controller.php]: This change breaks the experience in legacy checkout settings. To reproduce-
- Enable legacy checkout
- Go to the payment methods tab. Refresh the page.
- Notice that the card method is disabled.
- Enable the card method, save changes and refresh the page.
- Notice that the card method is disabled again.

When UPE is disabled we should return `WC_Stripe_Helper::get_legacy_enabled_payment_method_ids()` as before.

File: includes/admin/class-wc-rest-stripe-settings-controller.php [hunk 6]
```diff
@@ -24,6 +24,13 @@ class WC_REST_Stripe_Settings_Controller extends WC_Stripe_REST_Base_Controller
 	 */
 	private $gateway;
 
+	/**
+	 * Stripe payment gateway.
+	 *
+	 * @var WC_Stripe_Payment_Method_Configurations
+	 */
+	private $payment_method_configurations;
```
Comment (Mayisha) [inline includes/admin/class-wc-rest-stripe-settings-controller.php]: QQ: this is unused at this moment. Is it declared for future PRs?

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 2 above)
Comment (hsingyuc) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: @diegocurbelo I think that makes sense.  One thing to keep in mind is that we will also need to fetch the stripe payment configuration on every checkout page because we need to check for enabled payment methods there too.

Is that reasonable or should we implement non-admin only caching?

File: includes/payment-methods/class-wc-stripe-upe-payment-method-amazon-pay.php [hunk 7]
```diff
@@ -42,7 +42,8 @@ public static function is_amazon_pay_enabled() {
 			return false;
 		}
 
-		$upe_enabled_method_ids = WC_Stripe_Helper::get_settings( null, 'upe_checkout_experience_accepted_payments' );
+		$wc_stripe_upe_payment_method_amazon_pay = new self();
+		$upe_enabled_method_ids = $wc_stripe_upe_payment_method_amazon_pay->get_upe_enabled_payment_method_ids();
```
Comment (Mayisha) [inline includes/payment-methods/class-wc-stripe-upe-payment-method-amazon-pay.php:77]: How about calling the parent method `parent::update_payment_method_configurations()`?

File: includes/payment-methods/class-wc-stripe-upe-payment-method.php [hunk 8]
```diff
@@ -126,13 +126,14 @@ public function __construct() {
...
 
-		$this->enabled                  = $is_stripe_enabled && in_array( static::STRIPE_ID, $this->get_option( 'upe_checkout_experience_accepted_payments', [ WC_Stripe_Payment_Methods::CARD ] ), true ) ? 'yes' : 'no'; // @phpstan-ignore-line (STRIPE_ID is defined in classes using this class)
+		$this->enabled                  = $is_stripe_enabled && in_array( static::STRIPE_ID, $this->get_upe_enabled_payment_method_ids(), true ) ? 'yes' : 'no'; // @phpstan-ignore-line (STRIPE_ID is defined in classes using this class)
 		$this->id                       = WC_Gateway_Stripe::ID . '_' . static::STRIPE_ID; // @phpstan-ignore-line (STRIPE_ID is defined in classes using this class)
 		$this->has_fields               = true;
 		$this->testmode                 = WC_Stripe_Mode::is_test();
 		$this->supports                 = [ 'products', 'refunds' ];
 		$this->supports_deferred_intent = true;
 		$this->spe_enabled              = WC_Stripe_Feature_Flags::is_spe_available() && 'yes' === $this->get_option( 'single_payment_element' );
+		$this->stripe_api               = WC_Stripe_API::get_instance();
```
Comment (Mayisha) [inline includes/payment-methods/class-wc-stripe-upe-payment-method.php]: QQ: Is it declared for future use?

File: includes/class-wc-stripe-payment-method-configurations.php [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+					$payment_method_status = 'on' === $payment_method->display_preference->value;
+
+					if ( $payment_method_status ) {
+						$enabled_payment_method_ids[] = $payment_method_id;
+					}
+				}
+			}
+		}
+
+		set_transient( 'woocommerce_stripe_upe_enabled_payment_method_ids', $enabled_payment_method_ids, DAY_IN_SECONDS );
```
Comment (Mayisha) [inline includes/class-wc-stripe-payment-method-configurations.php]: We cache the [account response data for 2 hours](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/trunk/includes/class-wc-stripe-account.php#L132). We can use the same expiration time here.

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 2 above)
Comment (diegocurbelo) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: Non-admin only caching is a great idea 👍 … the checkout can potentially generate a lot of requests to the configurations api. And we can update that cache when we fetch new data in the admin pages.

**Comment by Mayisha** [comment]
I see this fatal error in my logs. It is probably from the time when I was testing the card enable/disable legacy experience.

Screenshot 2025-03-27 at 2 38 46 AM

File: includes/class-wc-stripe-payment-method-configurations.php (hunk 9 above)
Comment (Mayisha) [reply includes/class-wc-stripe-payment-method-configurations.php]: On second thought, 2 hours also seems a bit long. When I enable any method on Stripe dashboard, it does not update in `wp-admin` because of the caching.

Unfortunately, Stripe does not trigger any webhooks when we enable/disable payment methods on Stripe dashboard.

File: includes/payment-methods/class-wc-stripe-upe-payment-method-amazon-pay.php (hunk 7 above)
Comment (hsingyuc) [reply includes/payment-methods/class-wc-stripe-upe-payment-method-amazon-pay.php:77]: Since `get_upe_enabled_payment_method_ids` is not a static function, we need to call from an instance and the parent is abstract.  Do you have something else in mind?

File: includes/payment-methods/class-wc-stripe-upe-payment-method.php (hunk 8 above)
Comment (hsingyuc) [reply includes/payment-methods/class-wc-stripe-upe-payment-method.php]: Thank you! I forgot to remove it.

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 6 above)
Comment (hsingyuc) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: Thank you! I forgot to remove it.

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 5 above)
Comment (hsingyuc) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: Thank you! I forgot to add it back after testing.

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 4 above)
Comment (hsingyuc) [reply includes/admin/class-wc-rest-stripe-settings-controller.php:642]: It's removed in favor of #3920.

File: includes/class-wc-stripe-payment-method-configurations.php (hunk 9 above)
Comment (hsingyuc) [reply includes/class-wc-stripe-payment-method-configurations.php]: I removed the transient in this PR to meet the expectation. We can change it later if needed.

**Comment by hsingyuc** [comment]
@woocommerce/quark The PR is ready for review. Thank you!

File: includes/class-wc-stripe-payment-method-configurations.php [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+			foreach ( $merchant_payment_method_configuration as $payment_method_id => $payment_method ) {
+				if ( is_object( $payment_method ) &&
+					property_exists( $payment_method, 'display_preference' ) &&
+					property_exists( $payment_method->display_preference, 'value' ) ) {
+
+					$payment_method_status = 'on' === $payment_method->display_preference->value;
+
+					if ( $payment_method_status ) {
+						$enabled_payment_method_ids[] = $payment_method_id;
+					}
```
Comment (annemirasol) [inline includes/class-wc-stripe-payment-method-configurations.php]: Nit: You can use `isset()` to safely check for properties several levels deep. It will bail early at the first null encountered.

```
if ( isset( $payment_method->display_preference->value ) && 
    'on' === $payment_method->display_preference->value ) {
    $enabled_payment_method_ids[] = $payment_method_id;
}
```

File: includes/admin/stripe-settings.php [hunk 11]
```diff
@@ -307,14 +307,6 @@
...
 		],
 	];
-	if ( WC_Stripe_Feature_Flags::is_upe_checkout_enabled() ) {
-		// This adds the payment method section
-		$upe_settings['upe_checkout_experience_accepted_payments'] = [
-			'title'   => __( 'Payments accepted on checkout (Early access)', 'woocommerce-gateway-stripe' ),
-			'type'    => 'upe_checkout_experience_accepted_payments',
-			'default' => [ WC_Stripe_Payment_Methods::CARD, WC_Stripe_Payment_Methods::LINK ],
-		];
-	}
```
Comment (annemirasol) [inline includes/admin/stripe-settings.php:317]: 🧹 ✨

File: includes/admin/class-wc-rest-stripe-settings-controller.php [hunk 12]
```diff
@@ -722,7 +707,7 @@ public function dismiss_customization_notice( WP_REST_Request $request ) {
 		return new WP_REST_Response( [ 'result' => 'notice dismissed' ], 200 );
 	}
 
-	/**
+		/**
```
Comment (annemirasol) [inline includes/admin/class-wc-rest-stripe-settings-controller.php]: Unintended whitespace change.

File: includes/class-wc-stripe-payment-method-configurations.php [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+	/**
+	 * Get the merchant payment method configuration in Stripe.
+	 *
+	 * @return object|null
+	*/
+	private static function get_primary_configuration() {
+		if ( null !== self::$primary_configuration ) {
+			return self::$primary_configuration;
+		}
```
Comment (annemirasol) [inline includes/class-wc-stripe-payment-method-configurations.php:47]: :chefs-kiss: I like that we've pivoted to using this caching pattern, instead of the previous transient. While we do end up with more API calls, it avoids the unidirectional update problem we had with transients, e.g. if merchant updates via Stripe dashboard, we have no way of invalidating our transient.

File: includes/class-wc-stripe-payment-method-configurations.php [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+		$result = WC_Stripe_API::get_instance()->get_payment_method_configurations();
+		$payment_method_configurations = $result->data ?? null;
+
+		if ( ! $payment_method_configurations ) {
+			return null;
+		}
+
+		foreach ( $payment_method_configurations as $payment_method_configuration ) {
+			if ( ! $payment_method_configuration->livemode && $payment_method_configuration->parent && 'pmc_1LEKjBGX8lmJQndTBOzjqxSa' === $payment_method_configuration->parent ) {
```
Comment (annemirasol) [inline includes/class-wc-stripe-payment-method-configurations.php]: What do we think about moving the config ID string, and the one below, to a constant?

**Review by annemirasol — COMMENTED** [review]
👏 

Tested:
✅  disabling/enabling payment methods in wp-admin and verifying consistency with Stripe dashboard
✅ disabling/enabling payment methods in Stripe dashboard and verifying consistency with wp-admin
✅ disabling/enabling payment methods in wp-admin and verifying consistency with block/classic checkout
✅ Enabling an "unsupported currency" payment method in the Stripe dashboard, and verifying it remains disabled in wp-admin and block/classic checkout

Some side notes:
- Will we be having a follow-up PR to migrate merchant settings? 
- It looks like we make 13 (?) calls to `GET payment_method_configurations` when loading the settings page. I understand we can't use transients, and this is something we can possibly improve on later.

I think this is good to go, although I do see some phpunit tests failing (ran locally). Do you know if those are related?

**Comment by hsingyuc** [comment]
> Will we be having a follow-up PR to migrate merchant settings?

Yes, that would be #4118.  

I made your other changes.  Let me know what you think!

**Review by annemirasol — CHANGES_REQUESTED** [review]
On the retest, found this issue:
- When turning off legacy checkout using the checkbox (not the banner), we get a fatal:
```
PHP Fatal error:  Uncaught Error: Call to undefined method WC_Gateway_Stripe::update_enabled_payment_methods() in /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/admin/class-wc-rest-stripe-settings-controller.php:627
Stack trace:
#0 /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/admin/class-wc-rest-stripe-settings-controller.php(298): WC_REST_Stripe_Settings_Controller->update_enabled_payment_methods(Object(WP_REST_Request))
#1 /var/www/html/wp-includes/rest-api/class-wp-rest-server.php(1292): WC_REST_Stripe_Settings_Controller->update_settings(Object(WP_REST_Request))
#2 /var/www/html/wp-includes/rest-api/class-wp-rest-server.php(1125): WP_REST_Server->respond_to_request(Object(WP_REST_Request), '/wc/v3/wc_strip...', Array, NULL)
#3 /var/www/html/wp-includes/rest-api/class-wp-rest-server.php(439): WP_REST_Server->dispatch(Object(WP_REST_Request))
#4 /var/www/html/wp-includes/rest-api.php(449): WP_REST_Server->serve_request('/wc/v3/wc_strip...')
#5 /var/www/html/wp-includes/class-wp-h in /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/admin/class-wc-rest-stripe-settings-controller.php on line 627
```
Steps to reproduce:
- Enable legacy checkout.
- Disable legacy checkout, using the checkbox (not the banner).
- See fatal in logs, and "Error saving settings" error message.



Retested, this time on EUR:
✅ disabling/enabling payment methods in wp-admin and verifying consistency with Stripe dashboard
✅ disabling/enabling payment methods in Stripe dashboard and verifying consistency with wp-admin
✅ disabling/enabling payment methods in wp-admin and verifying consistency with block/classic checkout
✅ enabling an "unsupported currency" payment method in the Stripe dashboard, and verifying it remains disabled in wp-admin and block/classic checkout

**Comment by hsingyuc** [comment]
@annemirasol, good find! Fixed [aa4066c](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4091/commits/aa4066ca3af686fc434352062a2d76b6c435aded). I just realized I got let go, can you take over from there. Thank you.

**Comment by annemirasol** [comment]
✅ Retested consistency between the Stripe dashboard, wp-admin and store front.

As we are merging to a feature branch and not `develop`, I'm okay with merging this as is to unblock other work, and just do follow-up PRs if needed for the following issues I've found:

- e2e tests are repeatedly failing for the `enable Link` setup step. (This is not failing in my local runs.)
- Edge case (and possibly not a realistic scenario): if you uncheck every payment method in wp-admin and click "Save changes", we get a "Settings saved" confirmation, but changes were not actually saved.
- Is this expected? Disconnecting your Stripe account will turn off certain payment methods, like Affirm, Afterpay and CashApp.
- When we turn on a currency-incompatible payment method in the dashboard, the Payment Methods settings page and store front do not display them as enabled (expected), but other wp-admin pages display the `Xyz is enabled - it requires store currency to be set to EUR` warning message (see screenshot).

Screenshot 2025-04-07 at 4 15 14 PM

Edited to add: we're also displaying an incorrect count:

Screenshot 2025-04-07 at 4 43 10 PM

**Review by annemirasol — APPROVED** [review]

**Comment by diegocurbelo** [comment]
> ✅ Retested consistency between the Stripe dashboard, wp-admin and store front.
> 
> As we are merging to a feature branch and not `develop`, I'm okay with merging this as is to unblock other work, and just do follow-up PRs if needed for the following issues I've found:
> ...

Thank you @annemirasol; I'll add the issues as known issues to the `settings-sync` PR.
